### PR TITLE
Pregenerate RSA keys and speed up bootstrap times

### LINKED ISSUE #1203: Pregenerate RSA keys and speed up bootstrap times
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**
The operator CPU cycles are limited. It often sits almost idle but when a new ScyllaCluster is created (or other CRs needing certs), it needs to generate a few RSA keys for certificates to represent the trust chains which often hits the CPU limit and takes a while. We can create a buffer of pre-generated keys that will make sure the CPU load is better spread over time and in the ideal case the keys are available right when needed.

**What is use case behind this feature:**
UX, better utilization

**Additional Information:**

### PR DESCRIPTION
**Pull request by tnozicka**
**Description of your changes:**
The operator CPU cycles are limited. It often sits almost idle but when a new ScyllaCluster is created (or other CRs needing certs), it needs to generate a few RSA keys for certificates to represent the trust chains which often hits the CPU limit and takes a while. This PR adds a buffer of pre-generated keys that will make sure the CPU load is better spread over time and in the ideal case the keys are available right when needed.

**Which issue is resolved by this Pull Request:**
Resolves #1203

### CODE REVIEW

File: pkg/cmd/operator/operator.go [hunk 1]
```diff
@@ -110,6 +116,14 @@ func (o *OperatorOptions) Validate() error {
 		errs = append(errs, errors.New("operator image can't be empty"))
 	}
 
+	if len(o.OperatorImage) == 0 {
+		errs = append(errs, errors.New("operator image can't be empty"))
+	}
+
+	if o.CryptoKeyBufferSize < 1 {
+		errs = append(errs, errors.New("crypto buffer key size has to be at least 1"))
```
Comment (mykaul) [inline pkg/cmd/operator/operator.go:124]:
```suggestion
		errs = append(errs, errors.New("crypto-key-buffer-size has to be at least 1"))
```

File: hack/ci-deploy.sh [hunk 2]
```diff
@@ -45,6 +45,7 @@ for f in $( find "${deploy_dir}"/ -type f -name '*.yaml' ); do
 done
 
 yq e --inplace '.spec.template.spec.containers[0].args += ["--qps=200", "--burst=400"]' "${deploy_dir}/operator/50_operator.deployment.yaml"
+yq e --inplace '.spec.template.spec.containers[0].args += ["--crypto-key-buffer-size-min=3", "--crypto-key-buffer-size-max=6", "--crypto-key-buffer-delay=2s"]' "${deploy_dir}/operator/50_operator.deployment.yaml"
```
Comment (zimnx) [inline hack/ci-deploy.sh:48]: are default numbers causing any issues in CI?

Comment (tnozicka) [reply hack/ci-deploy.sh:48]: We are very limited on CPU cycles and we do not yet have limits on operator consumption (#1204) so as the cpu is shared for all processes any extra work will extend the CI run time at best, possibly times out the suite it individual waits in it. I've tried it when the default were higher and it didn't go well.

File: pkg/cmd/operator/operator.go [hunk 3]
```diff
@@ -32,6 +34,10 @@ import (
 	"k8s.io/klog/v2"
 )
 
+var (
```
[the remark this answers is not in the document]
Comment (tnozicka) [reply pkg/cmd/operator/operator.go]: fixed

File: pkg/itemgenerator/generator.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+	case <-ctx.Done():
+		return ctx.Err()
+
+	case ch <- item:
+		return nil
+	}
+}
+
+func (g *Generator[T]) runUnlimitedWorker(ctx context.Context) {
+	klog.V(4).InfoS("Interface is starting unlimited worker", "Name", g.name)
```
[the remark this answers is not in the document]
Comment (tnozicka) [reply pkg/itemgenerator/generator.go]: fixed

File: pkg/itemgenerator/generator.go [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		g.runRateLimitedWorker(ctx)
+	}()
+
+	defer func() {
+		klog.V(4).InfoS("Interface workers have finished.", "Name", g.name, "ContextError", ctx.Err())
```
[the remark this answers is not in the document]
Comment (tnozicka) [reply pkg/itemgenerator/generator.go]: fixed the defer order, thx

File: pkg/itemgenerator/generator.go [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+}
+
+func (g *Generator[T]) runWorker(ctx context.Context, ch chan<- *T) error {
+	klog.V(4).InfoS("Generating item", "Name", g.name, "BufferCapacity", cap(ch), "BufferSize", len(ch))
+
+	item, err := g.generateFunc()
+	if err != nil {
+		return fmt.Errorf("generator %q: can't generate item: %w", g.name, err)
+	}
+	klog.V(4).InfoS("Generating item finished", "Name", g.name, "BufferCapacity", cap(ch), "BufferSize", len(ch))
```
[the remark this answers is not in the document]
Comment (tnozicka) [reply pkg/itemgenerator/generator.go]: good catch, thx

**Review by zimnx — APPROVED** [review]
LGTM, thanks

**Pull request merged**